### PR TITLE
New short-read assembler: VelvetOptimizer

### PR DESCRIPTION
**Pull request by danejo3**
In this PR, a new assembly algorithm called `VelvetOptimizer` has been added. `VelvetOptimizer` is a wrapper script for the original `Velvet` assembly algorithm. It calculates a variety of parameters to optimize accuracy and improve results for the final assembly. Internal tests have shown that `Velvet` is capable of assembling complex plant genomes.

`VelvetOptimizer` has three particularly important parameters that users should consider to YEAT's config `extra_args` parameter:

- -s (--hashs): The lower bound of the hash value range that the optimizer will search to find the optimal value.
- -e (--hashe): The upper bound of the hash value range that the optimizer will search to find the optimal value.
- -x (--step): The number of hash values to skip with each increment from -s to -e while searching for the optimal value.

The following values were particularly useful during our plant genome assembly runs:

- -s 77
- -e 127
- -x 10

To run `VelvetOptimizer`, users will need to create a separate conda environment dedicated to `Velvet`. This is because installing `Velvet` updates C files used by `Canu`, which causes `Canu` to fail during execution. The specific files responsible for this issue are unknown based on my observations. As long as the user has a conda environment named "Velvet", YEAT will execute `VelvetOptimizer` in a manner similar to YEAT's `metaMDBG` (OS-dependent) workflow.

**NOTE:**
In order to run `VelvetOptimizer` with YEAT, run the following line in the CLI.

```
conda create -n velvet -c bioconda "perl-velvetoptimiser>=2.2" -y
```

### CODE REVIEW

**Comment by danejo3** [comment]
```
rule velvet:
    output:
        contigs="analysis/{sample}/paired/{label}/velvet/contigs.fasta"
    input:
        read1="seq/downsample/{sample}/paired/{sample}.R1.fq.gz",
        read2="seq/downsample/{sample}/paired/{sample}.R2.fq.gz",
        genome_size="seq/downsample/{sample}/paired/genome_size.txt"
    threads: 128
    params:
        temp_dir="analysis/{sample}/paired/{label}/velvet-temp",
        actual_dir="analysis/{sample}/paired/{label}/velvet",
        extra_args=lambda wildcards: config["assemblies"][wildcards.label].extra_args,
        low=lambda wildcards: config["assemblies"][wildcards.label].low,
        high=lambda wildcards: config["assemblies"][wildcards.label].high,
        step=lambda wildcards: config["assemblies"][wildcards.label].step,
        reads=lambda wildcards, input: f"'-shortPaired -fastq.gz -separate {input.read1} {input.read2}'",
        
    run:
        genome_size = get_genome_size1(input.genome_size)
        shell("VelvetOptimiser.pl -s {params.low} -e {params.high} -x {params.step} -g {genome_size} -f {params.reads} -t {threads} -d {params.temp_dir} {params.extra_args}")
        shell("mv {params.temp_dir}/* {params.actual_dir}")
        shell("rm -r {params.temp_dir}")
        shell("ln -s contigs.fa {output.contigs}")
```

File: README.md [hunk 1]
```diff
@@ -7,7 +7,8 @@ YEAT, **Y**our **E**verday **A**ssembly **T**ool, is an update to [`asm_tools`](
 ```
 git clone https://github.com/bioforensics/yeat.git
 cd yeat
-conda env create --name yeat --file environment.yml
+conda env create -f environment.yml
+conda create -n velvet -c bioconda "perl-velvetoptimiser>=2.2" -y
```
Comment (danejo3) [inline README.md:11]: Quick instruction to install `VelvetOptimiser` in another conda enviroment.

File: yeat/tests/data/configs/paired.cfg [hunk 2]
```diff
@@ -33,6 +33,22 @@
...
             "mode": "paired"
+        },
+        "penguin-default": {
+            "algorithm": "penguin",
+            "extra_args": "",
+            "samples": [
+                "Shigella_sonnei_53G"
+            ],
+            "mode": "paired"
+        },
```
Comment (danejo3) [inline yeat/tests/data/configs/paired.cfg:44]: Test data config files for the `Penguin` assembler in #82 were not updated. Updating in this PR.

File: yeat/workflow/Bandage [hunk 3]
```diff
@@ -140,3 +140,18 @@ rule metamdbg:
...
         """
+
+
+rule velvet:
+    output:
+        status="analysis/{sample}/{readtype}/{label}/velvet/bandage/.done"
+    input:
+        report="analysis/{sample}/{readtype}/{label}/velvet/quast/report.html"
+    params:
+        gfa=lambda wildcards: glob(f"analysis/{wildcards.sample}/{wildcards.readtype}/{wildcards.label}/velvet/auto_data_*/Graph*"),
```
Comment (danejo3) [inline yeat/workflow/Bandage]: `VelvetOptimiser` produces only 2 GFA files for Bandage.

File: yeat/workflow/Paired [hunk 4]
```diff
@@ -78,3 +78,26 @@ rule penguin:
...
+        temp_dir="analysis/{sample}/paired/{label}/velvet-temp",
+        actual_dir="analysis/{sample}/paired/{label}/velvet",
+        extra_args=lambda wildcards: config["assemblies"][wildcards.label].extra_args,
+        reads=lambda wildcards, input: f"'-fastq.gz -shortPaired {input.read1} -shortPaired2 {input.read2}'",
+    shell:
+        """
+        VelvetOptimiser.pl -f {params.reads} -t {threads} -p {params.actual_dir}/auto -d {params.temp_dir} {params.extra_args}
+        mv {params.temp_dir}/* {params.actual_dir}
+        rm -r {params.temp_dir}
+        ln -s contigs.fa {output.contigs}
```
Comment (danejo3) [inline yeat/workflow/Paired:102]: VelvetOptimiser has a step where users can direct what location the final files are created at. If the output directory exists, VelvetOptimiser will fail. To resolve this issue, a "temp" directory was created for the final output and then the files within are then transferred to the intended output directory. (Similar approach to MEGAHIT was applied here.)

**Review by danejo3 — COMMENTED** [review]
This PR is ready for review! @standage 

Before running your own tests. Please run:

```
conda create -n velvet -c bioconda "perl-velvetoptimiser>=2.2" -y
```

The changes to the README and instructions for downloading VelvetOptimizer are quite sparse. I would like to propose a dedicated update in another PR to do a more thorough overhaul of the README file. I have a few ideas about what to add and remove, and I'll share those proposals there, if that's okay with you!

File: yeat/tests/data/configs/single.cfg [hunk 5]
```diff
@@ -33,6 +33,22 @@
...
+            "algorithm": "penguin",
+            "extra_args": "",
+            "samples": [
+                "Shigella_sonnei_53G"
+            ],
+            "mode": "single"
+        },
+        "velvet-default": {
+            "algorithm": "unicycler",
+            "extra_args": "",
```
Comment (standage) [inline yeat/tests/data/configs/single.cfg:47]: Typo: shouldn't this be `"algorithm": "velvet"`?

File: yeat/workflow/Paired [hunk 6]
```diff
@@ -78,3 +78,26 @@ rule penguin:
...
+
+
+rule velvet:
+    output:
+        contigs="analysis/{sample}/paired/{label}/velvet/contigs.fasta"
+    input:
+        read1="seq/downsample/{sample}/paired/{sample}.R1.fq.gz",
+        read2="seq/downsample/{sample}/paired/{sample}.R2.fq.gz"
+    conda:
+        "velvet"
```
Comment (standage) [inline yeat/workflow/Paired:90]: I'm going to reiterate my suggestion that this conda environment has a `velvet-` prefix to act as a kind of namespace to reduce the (admittedly already low) likelihood of a collision with an existing environment.

File: yeat/workflow/Paired (hunk 4 above)
Comment (standage) [reply yeat/workflow/Paired:102]: And this is a problem because Snakemake creates that output dir automatically, right? Good solution.

**Review by standage — CHANGES_REQUESTED** [review]
Looks good! Just a couple points to resolve.

File: yeat/tests/data/configs/single.cfg (hunk 5 above)
Comment (danejo3) [reply yeat/tests/data/configs/single.cfg]: Yes, good catch!

**Comment by danejo3** [comment]
I have one comment that I just wanted to make sure we were on the same page. Other than that, ready for another review! Thanks!

File: yeat/workflow/Paired [hunk 7]
```diff
@@ -78,3 +78,26 @@ rule penguin:
...
+
+
+rule velvet:
+    output:
+        contigs="analysis/{sample}/paired/{label}/velvet/contigs.fasta"
+    input:
+        read1="seq/downsample/{sample}/paired/{sample}.R1.fq.gz",
+        read2="seq/downsample/{sample}/paired/{sample}.R2.fq.gz"
+    conda:
+        "yeat-velvet"
```
[the remark this answers is not in the document]
Comment (standage) [reply yeat/workflow/Paired:90]: Yes, precisely.

[the remark this answers is not in the document]
Comment (danejo3) [reply yeat/workflow/Paired:90]: Great! We're good then with the recent push!

**Review by standage — APPROVED** [review]

**Comment by standage** [comment]
Tested successfully with `just-yeat-it` and the `short_reads_?.fastq.gz` data.

**Pull request merged**